src/software/ai/navigator/navigator.cpp

```diff
@@ -14,215 +15,159 @@ Navigator::Navigator(std::unique_ptr<PathManager> path_manager,
 
-    for (auto &mi_primitive :
-         getPrimitivesFromMoveIntents(move_intents_for_path_planning, world))
+    for (const auto &intent : intents)
     {
-        assigned_primitives.emplace_back(std::move(mi_primitive));
+        current_primitive = std::nullopt;
+        intent->accept(*this);
+        robot_primitives_map[intent->getRobotId()] = *current_primitive;
     }
-
-    return assigned_primitives;
+    return primitive_set_msg;
 }
 
-std::unordered_set<PathObjective> Navigator::getPathObjectivesFromMoveIntents(
-    const std::vector<MoveIntent> &move_intents, const World &world)
+std::vector<PathObjective> Navigator::createPathObjectives(
+    const std::vector<std::unique_ptr<Intent>> &intents, const World &world)
 {
-    std::unordered_set<PathObjective> path_objectives;
+    std::vector<ObstaclePtr> friendly_non_navigating_robot_obstacles;
+    std::vector<PathObjective> path_objectives;
+    const auto ball_obstacle =
+        robot_navigation_obstacle_factory.createFromBallPosition(world.ball().position());
 
-    for (const auto &intent : move_intents)
+    for (const auto &intent : intents)
     {
-        // start with non-MoveIntent robots and then add motion constraints
-        auto obstacles = friendly_non_move_intent_robot_obstacles;
-
-        auto motion_constraint_obstacles =
-            robot_navigation_obstacle_factory.createFromMotionConstraints(
-                intent.getMotionConstraints(), world);
-        obstacles.insert(obstacles.end(), motion_constraint_obstacles.begin(),
-                         motion_constraint_obstacles.end());
-
-        if (intent.getBallCollisionType() == BallCollisionType::AVOID)
-        {
-            auto ball_obstacle = robot_navigation_obstacle_factory.createFromBallPosition(
-                world.ball().position());
-            obstacles.push_back(ball_obstacle);
-        }
-
-        auto robot = world.friendlyTeam().getRobotById(intent.getRobotId());
-
+        unsigned int robot_id = intent->getRobotId();
+        auto robot            = world.friendlyTeam().getRobotById(robot_id);
         if (robot)
         {
-            Point start = robot->position();
-            Point end   = intent.getDestination();
+            std::vector<ObstaclePtr> obstacles =
+                robot_navigation_obstacle_factory.createFromMotionConstraints(
+                    intent->getMotionConstraints(), world);
+
+            if (intent->getBallCollisionType() == BallCollisionType::AVOID)
+            {
+                obstacles.push_back(ball_obstacle);
+            }
 
-            path_objectives.insert(PathObjective(start, end, robot->velocity().length(),
-                                                 obstacles, intent.getRobotId()));
+            Point path_obj_end;
+            if (intent->getNavigationDestination())
+            {
+                path_obj_end = *intent->getNavigationDestination();
```

Comment:
Possibly confusing that even "non-navigating intents" just generate degenerate paths that the visitor then ignores.